### PR TITLE
Various fixes for Mirror Image and Summon Elemental spells

### LINKED ISSUE #5443: Mirror Image - Bug that hides nearby creature
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

When you cast the "Mirror Image" spell on a (2-hex ?) creature that touches the right edge in combat (Pheonix in my video), the mirror creature appears, but hides the enemy creature just below (zombies in my video).

https://user-images.githubusercontent.com/104630245/169724258-0fced2ec-43ab-4741-b749-6e1ac10de9f8.mp4

### Save file

[POL_0367bug.zip](https://github.com/ihhub/fheroes2/files/8750580/POL_0367bug.zip)

### Additional info

In savegame, just attack enemy Hero "Ibn Fadlan" with Hero "Sarakin"

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
fix #5443

https://user-images.githubusercontent.com/32623900/169828162-8544a0c4-2445-4e1d-a8e4-c72797be60d7.mp4

---

Also this PR fixes similar issue with Mirror Image, but arising for a different reason:

https://user-images.githubusercontent.com/32623900/169828319-f94f78c6-fb2d-43a6-9a37-f7a493e93d3d.mp4

https://user-images.githubusercontent.com/32623900/169828663-bbcd910b-9fd9-494e-9766-f8316e9b20ed.mp4

In this PR this works properly:

https://user-images.githubusercontent.com/32623900/169828794-00ea6b05-56f7-42a8-9783-2e821b082633.mp4

https://user-images.githubusercontent.com/32623900/169828818-3524e394-d217-4374-a035-0bd45aaed916.mp4

---

Also it seems that currently AI may potentially summon elementals of different types, which is prohibited by the rules of the game. This PR also fixes that.

### CODE REVIEW

File: src/fheroes2/ai/normal/ai_normal_spell.cpp [hunk 1]
```diff
@@ -83,8 +83,9 @@ namespace AI
         };
 
         for ( const Spell & spell : allSpells ) {
-            if ( !_commander->HaveSpellPoints( spell ) || !spell.isCombat() || ( !spell.isDamage() && retreating ) )
+            if ( !spell.isCombat() || arena.isDisableCastSpell( spell ) || !_commander->CanCastSpell( spell ) || ( !spell.isDamage() && retreating ) ) {
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_spell.cpp:86]: A very tiny improvement: we could swap `!spell.isDamage() && retreating` as **retreating** is usually the first check we need.

File: src/fheroes2/battle/battle_arena.cpp [hunk 2]
```diff
@@ -980,61 +980,71 @@ bool Battle::Arena::isSpellcastDisabled() const
...
             return true;
         }
         else if ( spell.isSummon() ) {
             const Unit * elem = GetCurrentForce().FindMode( CAP_SUMMONELEM );
-            bool affect = true;
+            bool canSummon = true;
 
-            if ( elem )
+            if ( elem ) {
                 switch ( spell.GetID() ) {
```
Comment (ihhub) [inline src/fheroes2/battle/battle_arena.cpp:1015]: Should we do `Monster mons( spell );` as we've done in the code below?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , I put few comments in this pull request. Could you please check them?

File: src/fheroes2/heroes/heroes_base.cpp [hunk 3]
```diff
@@ -385,7 +385,7 @@ bool HeroBase::CanCastSpell( const Spell & spell, std::string * res /* = nullptr
 
     if ( !HaveSpell( spell ) ) {
         if ( res ) {
-            *res = _( "The spell is not found." );
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/heroes/heroes_base.cpp:388]: Hi @ihhub this function might be called for AI-controlled hero as well, so using an assertion here is a bit dangerous.

[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/heroes/heroes_base.cpp:388]: No-no. I mean this:
```
if ( !HaveSpell( spell ) ) {
        if ( res ) {
            // How is it possible that a human player tries to cast a spell which doesn't exit?
            assert( 0 );
        }
...
```

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@oleg-derevenetz , well done!

**Pull request merged**